frontend_tests/node_tests/emoji.js

```diff
@@ -0,0 +1,25 @@
+set_global('$', global.make_zjquery());
+set_global('page_params', {});
+set_global('upload_widget', {});
+
+add_dependencies({
+    emoji_codes: 'generated/emoji/emoji_codes.js',
+    emoji: 'js/emoji.js',
+});
+
+(function test_build_emoji_upload_widget() {
+    upload_widget.build_widget = function (
```

Comment:
You might want to double check that `upload_widget.build_widget` is called, just like how you did in the `Image` set in the `initialize` function.